K2Bridge/Visitors/DateRangeVisitor.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+namespace K2Bridge.Visitors
+{
+    using System.Globalization;
+    using System.Text;
+    using K2Bridge.Models.Request.Aggregations;
+    using K2Bridge.Models.Response;
+
+    /// <content>
+    /// A visitor for the <see cref="DateRangeAggregation"/> element.
+    /// </content>
+    internal partial class ElasticSearchDSLVisitor : IVisitor
+    {
+        /// <inheritdoc/>
+        public void Visit(DateRangeAggregation dateRangeAggregation)
+        {
+            Ensure.IsNotNull(dateRangeAggregation, nameof(RangeAggregation));
+            EnsureClause.StringIsNotNullOrEmpty(dateRangeAggregation.Metric, nameof(RangeAggregation.Metric));
+            EnsureClause.StringIsNotNullOrEmpty(dateRangeAggregation.Field, nameof(RangeAggregation.Field));
+
+            var queryStringBuilder = new StringBuilder();
+
+            // Part 1:
+            // _data | extend ['_range'] = pack_array("range1", "range2", "range3"), ['_range_value']=pack_array(expr1, expr2, expr3)
+
+            // Start of query, until first pack_array()
+            queryStringBuilder.Append($"_data | {KustoQLOperators.Extend} {EncodeKustoField(dateRangeAggregation.Key)} = {KustoQLOperators.PackArray}(");
+
+            // Insert range names
+            foreach (var range in dateRangeAggregation.Ranges)
+            {
+                range.Field = dateRangeAggregation.Field;
+                range.Accept(this);
+
+                queryStringBuilder.Append($"{range.BucketNameKustoQL},");
+            }
+
+            // Remove final comma
+            queryStringBuilder.Remove(queryStringBuilder.Length - 1, 1);
+
+            // Close the first pack_array() and start the second pack_array()
+            queryStringBuilder.Append($"), ['_range_value'] = {KustoQLOperators.PackArray}(");
+
+            // Insert range expressions
+            foreach (var range in dateRangeAggregation.Ranges)
+            {
+                if (string.IsNullOrEmpty(range.KustoQL))
+                {
+                    // This is then "open" range, -infinity to +infinity
```

Comment:
If needed, you can put elements of this section in an intermediate List<string>. Then use string.Join(',',myList) and put the final result in the stringBuilder. It avoids the additional coma.

Reply:
This might be overkill, I think it's not that bad to call `Remove()` once or twice :-)